### PR TITLE
nautilus: rgw: add "rgw-orphan-list" tool and "radosgw-admin bucket radoslist ..."

### PR DESCRIPTION
**Pull request by ivancich**
backport trackers:

* https://tracker.ceph.com/issues/44722
* https://tracker.ceph.com/issues/45488
* https://tracker.ceph.com/issues/46344
* https://tracker.ceph.com/issues/46556
* https://tracker.ceph.com/issues/46557

---

Adds a radosgw-admin subcommand and possibly a bucket name (if not, all buckets are assumed) and walks the bucket(s) listing(s) and the manifest (if it exists) for each object in the listing to generate the rados objects that represent the rgw objects in the bucket.

Also adds a tool named rgw-orphan-list that will produce a list in a local file of what appear to be rgw orphans.

tracker: https://tracker.ceph.com/issues/44722

---

To Do:

- [x] https://github.com/ceph/ceph/pull/35009 merged into master
- [x] https://github.com/ceph/ceph/pull/35009 cherry-picked into this PR
- [x] clarify missing commit - #34148 has 6 commits, this backport PR has only 5? ("That commit eaa8f66 on the master applies to some code refactoring that occurred between nautilus and master. It does not apply to nautilus.")
- [x] https://github.com/ceph/ceph/pull/35245 merged
- [x] https://github.com/ceph/ceph/pull/35245 cherry-picked into this PR
- [x] cherry-pick b1daf30d61989d0813ea73ac19f36b4de6abf13d (from https://github.com/ceph/ceph/pull/35736) into this PR
- [x] rebase to resolve PendingReleaseNotes conflict
- [x] cherry-pick #35966 into this PR
- [x] cherry-pick #35979 into this PR

### CODE REVIEW

**Comment by ivancich** [comment]
@tchaikov and @smithfarm : This PR adds two new executables -- rgw-orphan-list (script) and ceph-diff-sorted (binary). The cmake files have been modified, but I understand that other build files will need to be altered as well, and this is outside my knowledge.

What do I need to do w.r.t. this pr and/or the associated tracker(s)?

ALSO, this originated downstream and in an older version, so that's why the nautilus version is emerging first. I'm currently working on the mimic and octopus versions.

**Comment by ivancich** [comment]
@tchaikov and @smithfarm : So in speaking with @tserlin I think I made the correct changes to ceph.spec.in. So I'll add you both as reviewers.

File: ceph.spec.in [hunk 1]
```diff
@@ -1827,6 +1827,8 @@ fi
 %{_bindir}/radosgw-token
 %{_bindir}/radosgw-es
 %{_bindir}/radosgw-object-expirer
+%{_bindir}/rgw-orphan-list
+%{_bindir}/ceph-diff-sorted
```
Comment (tchaikov) [inline ceph.spec.in]: could you apply following patch
```diff
diff --git a/debian/radosgw.install b/debian/radosgw.install
index 303791694c..b2930879ad 100644
--- a/debian/radosgw.install
+++ b/debian/radosgw.install
@@ -1,7 +1,9 @@
 lib/systemd/system/ceph-radosgw*
+usr/bin/ceph-diff-sorted
 usr/bin/radosgw
 usr/bin/radosgw-es
 usr/bin/radosgw-object-expirer
 usr/bin/radosgw-token
+usr/bin/rgw-orphan-list
 usr/lib/libradosgw.so*
 usr/share/man/man8/radosgw.8
```
as i think we also can benefit from these enhancements if they can be included in debian packages. FWIW, i am trying to list them in the alphabetical order.

File: src/rgw/CMakeLists.txt [hunk 2]
```diff
@@ -397,3 +397,8 @@ if(WITH_TESTS)
     ceph_rgw_multiparser
     DESTINATION bin)
 endif(WITH_TESTS)
+
+configure_file(${CMAKE_SOURCE_DIR}/src/rgw/rgw-orphan-list.in
```
Comment (tchaikov) [inline src/rgw/CMakeLists.txt]: i don't think we need to prepare `rgw-orphan-list.in` as a `config_file` as it's a ready-to-use script without replacing the substitution variables in it with the variables in cmake file. instead, i'd suggest
```cmake
install(PROGRAMS rgw-orphan-list
  DESTINATION bin)
```

in which i assume that `src/rgw/rgw-orphan-list.in` is renamed to `src/rgw/rgw-orphan-list`.

File: src/rgw/rgw_rados.h [hunk 3]
```diff
@@ -2548,6 +2548,10 @@ class RGWMPObj {
     meta = "";
     upload_id = "";
   }
+  friend std::ostream& operator<<(std::ostream& out, const RGWMPObj& obj) {
+    return out << "RGWMPObj:{ prefix=\"" << obj.prefix <<
```
Comment (tchaikov) [inline src/rgw/rgw_rados.h]: nit, could put
```c++
return out << "RGWMPObj:{ prefix=" << std::quoted(obj.prefix)
     << ", meta=" << std::quoted(obj.meta) << " }";
```

File: ceph.spec.in (hunk 1 above)
Comment (smithfarm) [inline ceph.spec.in]: +1 for alphabetical order

**Comment by smithfarm** [comment]
@ivancich Can you add a sentence like the following to the commit message:

"This is not a cherry-pick from master because the feature was originally written for an older version of Ceph, and is being gradually forward-ported to newer and newer versions."

File: src/rgw/rgw_rados.h (hunk 3 above)
Comment (ivancich) [reply src/rgw/rgw_rados.h]: I did not know about std::quoted -- thanks!

Comment (ivancich) [reply src/rgw/rgw_rados.h]: After trying out std::quoted in a few places, it appears it only works if the parameter is essentially a string; if it is ostream'able it produces an error. Oh well.

Comment (tchaikov) [reply src/rgw/rgw_rados.h]: yeah, that's a pain. probably we could have our own `ceph::quoted()` . fortunately, `std::quoted` works in this case.

**Comment by tchaikov** [comment]
```
/home/jenkins-build/build/workspace/ceph-pull-requests/src/test/cli/radosgw-admin/help.t: failed
--- /home/jenkins-build/build/workspace/ceph-pull-requests/src/test/cli/radosgw-admin/help.t
+++ /home/jenkins-build/build/workspace/ceph-pull-requests/src/test/cli/radosgw-admin/help.t.err
@@ -29,6 +29,7 @@
     bucket rewrite             rewrite all objects in the specified bucket
     bucket sync disable        disable bucket sync
     bucket sync enable         enable bucket sync
+    bucket radoslist           list rados objects backing bucket's objects
     bi get                     retrieve bucket index object entries
     bi put                     store bucket index object entries
     bi list                    list raw bucket index entries
```

dismissed my approval because we need to update the cli test. =)

**Comment by ivancich** [comment]
> the packaging and cmake bits look good to me.
> 
> @ivancich Eric, i just have a question. are we going to have this change in master? if the answer is yes, i assume the reason why we want to forward-port the change instead the other way around is to get this change in production sooner.

Yes, it's going to master as well, and I think I invited your review on that PR as well.

The reason for the forward porting is that it came from downstream and nautilus was the easiest port to make first. Getting it to master took additional work.

**Comment by ivancich** [comment]
jenkins test make check

**Comment by tchaikov** [comment]
- http://pulpito.ceph.com/kchai-2020-03-28_05:53:06-rgw-nautilus-kefu2-testing-2020-03-27-2234-distro-basic-smithi/
- http://pulpito.ceph.com/kchai-2020-03-28_09:23:03-rgw-nautilus-kefu2-testing-2020-03-28-1555-distro-basic-smithi/
- http://pulpito.ceph.com/kchai-2020-03-28_10:18:54-rgw-nautilus-kefu2-testing-2020-03-28-1555-distro-basic-smithi/

**Comment by ivancich** [comment]
I'm working on docs and tests (starting with the master pr). Let's make this DNM until I integrate those here.

**Comment by ivancich** [comment]
@smithfarm The master PR merged. So this is now ready for further review and QA. If you would glance over it and make sure there are no clerical or process issues given its non-standard history, I'd appreciate it.

See: #34148

**Comment by smithfarm** [comment]
Also, the master PR https://github.com/ceph/ceph/pull/34148 has six commits while this nautilus backport has only five. This one is missing:

https://github.com/ceph/ceph/commit/eaa8f66ec0e045df161b219506906568345a0b83

@ivancich Can you explain in the PR description why it was omitted?

**Comment by smithfarm** [comment]
@ivancich If I'm not mistaken, #35009 can now be cherry-picked into this PR.

**Comment by ivancich** [comment]
> @ivancich If I'm not mistaken, #35009 can now be cherry-picked into this PR.

@smithfarm I've cherry-picked the commit. However I'm not certain it's right. 14.2.9 was tagged prior to this sequence of commits. Yet the PendingReleaseNotes file is headed by "14.2.9". I wonder if it should be headed by "14.2.10" or "14.3.x". Do you know how this is supposed to work?

**Comment by smithfarm** [comment]
Yes, the heading should be "14.2.10" ( there will never be a "14.3.x" - see https://docs.ceph.com/docs/master/releases/general/#understanding-the-release-cycle ) because "14.2.9" was a hotfix release. I would suggest to fix the heading right here in this PR.

(By the way, you'll need to rebase, too - there is a conflict.)

**Comment by ivancich** [comment]
> Yes, the heading should be "14.2.10" ( there will never be a "14.3.x" - see https://docs.ceph.com/docs/master/releases/general/#understanding-the-release-cycle ) because "14.2.9" was a hotfix release. I would suggest to fix the heading right here in this PR.
> 
> (By the way, you'll need to rebase, too - there is a conflict.)

Thanks for your help, @smithfarm. I fixed the heading and rebased.

**Comment by smithfarm** [comment]
@ivancich Thanks. One last thing: I noticed that the original PR has six commits, but you only backported five of them. This one was omitted:

    rgw: add const-correctness to rgw_obj_manifest and svc_tier_rados

Can you explain in the PR body why that one is not appropriate for nautilus?

**Comment by ivancich** [comment]
> @ivancich Thanks. One last thing: I noticed that the original PR has six commits, but you only backported five of them. This one was omitted:
> 
> ```
> rgw: add const-correctness to rgw_obj_manifest and svc_tier_rados
> ```
> 
> Can you explain in the PR body why that one is not appropriate for nautilus? Something like: "Commit [eaa8f66](https://github.com/ceph/ceph/commit/eaa8f66ec0e045df161b219506906568345a0b83) was omitted because it's not directly related to the feature being backported."

That commit (eaa8f66ec0e045df161b219506906568345a0b83) on the master applies to some code refactoring that occurred between nautilus and master. It does not apply to nautilus.

**Comment by smithfarm** [comment]
DNM pending completion of To Do list in the PR description

**Comment by ivancich** [comment]
> DNM pending completion of To Do list in the PR description

I cherry-picked that commit and therefore removed DNM. Thank you for helping to navigate through all this complexity, @smithfarm !

**Comment by smithfarm** [comment]
@ivancich DNM is back :-) 

First, it looks like https://github.com/ceph/ceph/pull/35736 is a follow-on fix for this feature, so it would make sense to cherry-pick it right here in this PR.

Second, I zeroed out PendingReleaseNotes so it no longer contains any ballast from previous nautilus releases. That makes the PendingReleaseNotes situation much clearer, but does unfortunately require all feature backport PRs (includiung this one) to be rebased because they now show a conflict.

1.

**Comment by ivancich** [comment]
@smithfarm I'm happy to combine master PRs into fewer backport PRs. I just got merged two other related master PRs -- https://github.com/ceph/ceph/pull/35966 and https://github.com/ceph/ceph/pull/35979. Would you like those combined as well?

**Comment by ivancich** [comment]
@smithfarm I've incorporated https://github.com/ceph/ceph/pull/35736 . I won't adjust DNM until you let me know about the other two master PRs.

As always, I appreciate your help!

**Comment by smithfarm** [comment]
@ivancich Yes, I am convinced that we're approaching this in the right way here. I removed "nautilus" from the Backport field in both new tracker tickets you created. (If a feature that is being backported is implemented by *n* PRs, it's always preferable (in my mind) to cherry-pick all *n* PRs to the stable branch via a single backport PR.)

So, if there are any more follow-on fixes for this feature, as long as this backport PR is open feel free to just cherry-pick 'em in.

I went ahead and removed "nautilus" from the "Backport" field in the tracker tickets associated with these new follow-on fixes.

**Comment by smithfarm** [comment]
@ivancich I updated the checklist at the bottom of the PR description. Feel free to check off those action items as you complete them.

**Comment by smithfarm** [comment]
Also, @ivancich it would be great if you could open an octopus PR and cherry-pick these follow-on fixes there. That way, the fixes will get cherry-picked in the correct order, with no conflicts, and the whole backport experience for this feature will be smoother for it.

**Comment by smithfarm** [comment]
> Also, @ivancich it would be great if you could open an octopus PR and cherry-pick these follow-on fixes into it. 

I went ahead and opened it: https://github.com/ceph/ceph/pull/36027

**Comment by smithfarm** [comment]
@ivancich Ping (users seem to be yearning for this one)

**Comment by ivancich** [comment]
> @ivancich Ping (users seem to be yearning for this one)

Thanks, @smithfarm. I'll get on it.

**Comment by ivancich** [comment]
@smithfarm Something doesn't look quite right. I updated https://github.com/ivancich/ceph-fork/tree/nautilus-rados-list but I do not see the changes reflected on this PR. I don't recall there being a time delay previously.

Nonetheless, I believe I completed the check-list.

Thank you for all your help!!!

**Comment by smithfarm** [comment]
@cbodley @mattbenjamin This is a big feature backport - 10 commits, 28 files changed. Can you give it your blessing for it to go into nautilus?

**Comment by smithfarm** [comment]
```
test_safe_to_destroy (tasks.mgr.dashboard.test_osd.OsdTest)
2020-07-15 17:45:44,014.014 INFO:__main__:----------------------------------------------------------------------
2020-07-15 17:45:44,014.014 INFO:__main__:Traceback (most recent call last):
2020-07-15 17:45:44,014.014 INFO:__main__:  File "/home/jenkins-build/build/workspace/ceph-dashboard-pr-backend/qa/tasks/mgr/dashboard/test_osd.py", line 110, in test_safe_to_destroy
2020-07-15 17:45:44,015.015 INFO:__main__:    'stored_pgs': [],
2020-07-15 17:45:44,015.015 INFO:__main__:  File "/home/jenkins-build/build/workspace/ceph-dashboard-pr-backend/qa/tasks/mgr/dashboard/helper.py", line 328, in assertJsonBody
2020-07-15 17:45:44,015.015 INFO:__main__:    self.assertEqual(body, data)
2020-07-15 17:45:44,015.015 INFO:__main__:AssertionError: {'is_[13 chars]oy': False, 'message': '[errno -11] 1 pgs have[40 chars]ons'} != {'is_[13 chars]oy': True, 'active': [], 'missing_stats': [], [37 chars]: []}
2020-07-15 17:45:44,015.015 INFO:__main__:+ {'active': [],
2020-07-15 17:45:44,015.015 INFO:__main__:- {'is_safe_to_destroy': False,
2020-07-15 17:45:44,015.015 INFO:__main__:? ^                      ^^^^
2020-07-15 17:45:44,015.015 INFO:__main__:
2020-07-15 17:45:44,015.015 INFO:__main__:+  'is_safe_to_destroy': True,
2020-07-15 17:45:44,015.015 INFO:__main__:? ^                      ^^^
2020-07-15 17:45:44,015.015 INFO:__main__:
2020-07-15 17:45:44,015.015 INFO:__main__:-  'message': '[errno -11] 1 pgs have unknown state; cannot draw any conclusions'}
2020-07-15 17:45:44,015.015 INFO:__main__:+  'missing_stats': [],
2020-07-15 17:45:44,015.015 INFO:__main__:+  'safe_to_destroy': [13],
2020-07-15 17:45:44,016.016 INFO:__main__:+  'stored_pgs': []}
2020-07-15 17:45:44,016.016 INFO:__main__:
2020-07-15 17:45:44,016.016 INFO:__main__:----------------------------------------------------------------------
2020-07-15 17:45:44,016.016 INFO:__main__:Ran 72 tests in 810.202s
2020-07-15 17:45:44,016.016 INFO:__main__:
2020-07-15 17:45:44,016.016 INFO:__main__:
```

**Comment by smithfarm** [comment]
jenkins test dashboard backend

**Comment by smithfarm** [comment]
I went ahead and added the "needs-qa" and "nautilus-batch-1" labels. That means the PR will be included in one of @yuriw 's nautilus integration runs and, since the PR is already approved by @tchaikov and myself, it will get merged if it passes tests and nobody objects between now and then.

**Comment by yuriw** [comment]
[wip-yuri4-testing-2020-07-16-1539-nautilus](https://trello.com/c/4C5omBDz/965-wip-yuri4-testing-2020-07-16-1539-nautilus)

**Pull request merged**